Throw invalid_tenant_configuration when changing assigned cluster

sfc-gh-yajin opened the pull request:
Since we currently do not support tenant movement, we should as well explicitly disallow changing the assigned
cluster of a tenant during configuration by throwing `invalid_tenant_configuration` for now.

Test plan:
- add coverage for changing assigned cluster during tenant configuration
- fdbcli
- simulation tests

# Code-Reviewer Section

The general pull request guidelines can be found [here](https://github.com/apple/foundationdb/wiki/FoundationDB-Commit-Process).

Please check each of the following things and check *all* boxes before accepting a PR.

- [ ] The PR has a description, explaining both the problem and the solution.
- [ ] The description mentions which forms of testing were done and the testing seems reasonable.
- [ ] Every function/class/actor that was touched is reasonably well documented.

## For Release-Branches

If this PR is made against a release-branch, please also check the following:

- [ ] This change/bugfix is a cherry-pick from the next younger branch (younger `release-branch` or `main` if this is the youngest branch)
- [ ] There is a good reason why this PR needs to go into a release branch and this reason is documented (either in the description above or in a linked GitHub issue)

sfc-gh-ajbeamon commented on fdbclient/include/fdbclient/MetaclusterManagement.actor.h:
```diff
@@ -1749,6 +1749,23 @@ struct ConfigureTenantImpl {
+					throw invalid_tenant_configuration();
+				}
+			}
+		}
```
We have a loop below where we are processing the parameters (see line 1775), I think it would be reasonable to check for this condition there and avoid having to search them twice.

sfc-gh-yajin replied:
The motivation here is to return early to avoid unnecessary work, but I guess I can move the check there.

sfc-gh-yajin replied:
Done

sfc-gh-ajbeamon commented on fdbserver/workloads/TenantManagementWorkload.actor.cpp:
```diff
@@ -1512,6 +1512,13 @@ struct TenantManagementWorkload : TestWorkload {
+			if (deterministicRandom()->coinflip()) {
+				newClusterName = self->dataClusterName;
+			}
+			configuration["assigned_cluster"_sr] = newClusterName;
```
My guess is that these test updates are going to trigger some errors, so it may be necessary to check for those conditions. And in fact, we should probably ASSERT that we get an error if we know we're using a bad configuration.

sfc-gh-yajin replied:
Done

sfc-gh-yajin commented on fdbclient/include/fdbclient/TenantManagement.actor.h:
```diff
@@ -431,6 +431,19 @@ Future<Void> configureTenantTransaction(Transaction tr,
+			    .detail("NewAssignedCluster", updatedTenantEntry.assignedCluster.get().toString().c_str());
+			throw invalid_tenant_configuration();
+		}
+	}
```
SGTM.

sfc-gh-ajbeamon commented on fdbclient/include/fdbclient/MetaclusterManagement.actor.h:
```diff
@@ -1759,6 +1759,15 @@ struct ConfigureTenantImpl {
+				auto& newClusterName = configItr->second;
+				TraceEvent("UpdateManagementCluster")
+				    .detail("OriginalAssignedCluster",
+				            tenantEntry.get().assignedCluster.present() ? tenantEntry.get().assignedCluster.get()
+				                                                        : "null"_sr)
```
You can log an optional directly and it will print an appropriate string if the optional is not set:

```suggestion
				            tenantEntry.get().assignedCluster)
```

sfc-gh-ajbeamon commented on fdbclient/include/fdbclient/MetaclusterManagement.actor.h:
```diff
@@ -1759,6 +1759,15 @@ struct ConfigureTenantImpl {
+				    .detail("OriginalAssignedCluster",
+				            tenantEntry.get().assignedCluster.present() ? tenantEntry.get().assignedCluster.get()
+				                                                        : "null"_sr)
+				    .detail("NewAssignedCluster", newClusterName.present() ? newClusterName.get() : "null"_sr);
```
```suggestion
				    .detail("NewAssignedCluster", newClusterName);
```

sfc-gh-ajbeamon commented on fdbclient/include/fdbclient/MetaclusterManagement.actor.h:
```diff
@@ -1759,6 +1759,15 @@ struct ConfigureTenantImpl {
+			} else if (configItr->first == "assigned_cluster"_sr &&
+			           configItr->second != tenantEntry.get().assignedCluster) {
+				auto& newClusterName = configItr->second;
+				TraceEvent("UpdateManagementCluster")
```
```suggestion
				TraceEvent(SevWarn, "CannotChangeAssignedCluster")
```

sfc-gh-ajbeamon replied:
It might also be helpful to include the tenant name as a detail

sfc-gh-ajbeamon replied:
Can we also include the above suggestion to change the trace event type name and severity?

sfc-gh-ajbeamon commented on fdbserver/workloads/TenantManagementWorkload.actor.cpp:
```diff
@@ -1550,7 +1559,7 @@ struct TenantManagementWorkload : TestWorkload {
 					return Void();
 				} else if (e.code() == error_code_invalid_tenant_configuration) {
-					ASSERT(hasInvalidOption);
+					ASSERT(hasInvalidOption || configuration["assigned_cluster"_sr] == "newcluster"_sr);
```
We should also assert above (around L1534) that we aren't changing the cluster name. You could store a boolean at the start to track whether you think you are changing the cluster and then use that in both asserts to follow the convention of the other checks

sfc-gh-ajbeamon commented on fdbserver/workloads/TenantManagementWorkload.actor.cpp:
```diff
@@ -1512,6 +1512,18 @@ struct TenantManagementWorkload : TestWorkload {
+		// Configuring 'assignedCluster' requires reading existing tenant entry which is handled in
+		// `updateManagementCluster`.
+		state bool assignToDifferentCluster = false;
+		if (operationType == OperationType::METACLUSTER && (!hasInvalidOption || deterministicRandom()->coinflip())) {
```
The previous use of `!hasInvalidOption` was to make sure we did some configuration. In other words, we either used an invalid option or we set the tenant group.

Now, we have three candidates, so I might structure this a little bit differently to try all combinations.

One option is to coinflip both tenant group and assigned cluster, and then set an invalid option if there is no other configuration set or by coinflip. Something like this:

```
if (coinflip()) { setTenantGroup(); }
if (coinflip()) { setAsssignedCluster(); }
if (configuration.empty() || coinflip()) { setInvalidOption(); }
```

sfc-gh-yajin replied:
Maybe I am missing something. When looking at the code, if `operationType == SPECIAL_KEYS`,  how does `ConfigureTenantImpl` throw if the tenant does not exist?

sfc-gh-yajin replied:
Also just realize that in the case of SPECIAL_KEYS, we need to make sure `configuration` is not empty.

sfc-gh-ajbeamon reviewed (commented):
Overall this looks good, I have just two minor comments

sfc-gh-yajin commented:
Thanks @sfc-gh-ajbeamon for the review. I am also planning to update fdbcli_tests.py and metacluster_fdbcli_tests.py.

sfc-gh-yajin commented:
Thanks @sfc-gh-ajbeamon for the detailed review and help with investigation of joshua test failures. I have addressed the review comments and updated the PR.

sfc-gh-ajbeamon commented on fdbserver/workloads/TenantManagementWorkload.actor.cpp:
```diff
@@ -1566,11 +1565,9 @@ struct TenantManagementWorkload : TestWorkload {
 				ASSERT(!assignToDifferentCluster);
 				Versionstamp currentVersionstamp = wait(getLastTenantModification(self, operationType));
-				if (configurationChanging) {
+				if (tenantGroupChanging) {
```
We need the versionstamp check to be done if we are updating any configuration. Right now, `tenantGroupChanging` == `configurationChanging`, but that may not always be the case.

sfc-gh-yajin replied:
Make sense. Will update.

sfc-gh-jfu replied:
Maybe I'm misinterpreting this, but doesn't this still have the same issue? If we update line 1539 in the future, this line is checking the wrong configuration flag because it does not encode the possibility of `tenantGroupChanging || assignToDifferentCluster.` Is there any reason not to use `configurationChanging` all the way throughout instead of temporarily changing what we call it?

sfc-gh-yajin replied:
I think this block is specific to tenant-group changing, that's why it is inside the `if (tenantGroupChanging)`

sfc-gh-jfu replied:
Ok, I see what's going on now. The original comment was on an old revision and shows up on a different line number when I view it on the latest commit. I can see that this comment was referring to what is now line 1559. Sorry for the confusion

sfc-gh-yajin commented on fdbserver/workloads/TenantManagementWorkload.actor.cpp:
```diff
@@ -1524,14 +1536,16 @@ struct TenantManagementWorkload : TestWorkload {
+			if (exists &&
+			    (!itr->second.assignedCluster.present() || itr->second.assignedCluster.get() != newClusterName)) {
+				assignToDifferentCluster = true;
+			}
```
okay, I can remove it. Maybe we will need it in the future? we can add it back then.

sfc-gh-jfu commented on fdbcli/tests/metacluster_fdbcli_tests.py:
```diff
@@ -42,51 +88,75 @@ def metacluster_create(cluster_file, name, tenant_id_prefix):
   allocated tenant groups: 0
 """
-    expected = expected.format(num_clusters - 1).strip()
+    expected = expected.format(num_clusters - 1, 12).strip()
```
Where does this 12 come from? It's hard to follow the context here

sfc-gh-yajin replied:
It's because there are two data clusters each of which has `max_tenant_group=6`.

sfc-gh-yajin replied:
I can update the test script here if the remaining CI fails because I will have to re-trigger them anyway. :)

sfc-gh-jfu replied:
I see, can this be parameterized and passed around, or at least have a comment here describing how it's calculated? This seems like it would be hard to debug if we changed one or the other since there's nothing explicitly linking the two unless you already know or walk through the entire thing once slowly

sfc-gh-jfu replied:
Looks like CI has passed and this is pretty minor, so let's avoid kicking off CI again. You can either make another PR for this change, or I can apply it to my PR while I fix merge conflicts, if that's fine with you.

sfc-gh-yajin replied:
Sure, I will submit another PR to update the metacluster_fdbcli_tests.py.
Thanks

sfc-gh-yajin replied:
Addressed in #9406
